packages/vm/src/modules/file_system_cache.rs

```diff
@@ -14,7 +14,23 @@ use crate::errors::{VmError, VmResult};
 /// The string is used as a folder and should be named in a way that is
 /// easy to interprete for system admins. It should allow easy clearing
 /// of old versions.
-const MODULE_SERIALIZATION_VERSION: &str = "v1";
+///
+/// See https://github.com/wasmerio/wasmer/issues/2781 for more information
+/// on Wasmer's module stability concept.
+///
+/// ## Version history:
+/// - **v1**:<br>
+///   cosmwasm_vm < 1.0.0-beta5. This is working well up to Wasmer 2.0.0 as
+///   [in wasmvm 1.0.0-beta2](https://github.com/CosmWasm/wasmvm/blob/v1.0.0-beta2/libwasmvm/Cargo.lock#L1412-L1413)
+///   and [wasmvm 0.16.3](https://github.com/CosmWasm/wasmvm/blob/v0.16.3/libwasmvm/Cargo.lock#L1408-L1409).
+///   Versions that ship with Wasmer 2.1.x such [as wasmvm 1.0.0-beta3](https://github.com/CosmWasm/wasmvm/blob/v1.0.0-beta3/libwasmvm/Cargo.lock#L1534-L1535)
+///   to [wasmvm 1.0.0-beta5](https://github.com/CosmWasm/wasmvm/blob/v1.0.0-beta5/libwasmvm/Cargo.lock#L1530-L1531)
+///   are broken, i.e. they will crash when reading older v1 modules.
+/// - **v2**:<br>
+///   Version for cosmwasm_vm 1.0.0-beta5 / wasmvm 1.0.0-beta6 that ships with Wasmer 2.1.1.
+/// - **v3**:<br>
+///   Version for Wasmer 2.2.0 which contains a [module breaking change to 2.1.x](https://github.com/wasmerio/wasmer/pull/2747).
+const MODULE_SERIALIZATION_VERSION: &str = "v2";
```

Comment:
Can't this be derived or imported from wasmer somehow? That will make this code support all versions transparently (new test below would have to me modified).

Reply:
No, the version is not exposed yet. This might change in the future. They added a versioning schema in https://github.com/wasmerio/wasmer/pull/2747. However, it is not exported. So currently they just say it can change at any time.

If we get the version, we can check much better that we don't miss an update. However, we might want to keep control over the version number since it can include our own format changes as well. E.g. we could rename the file names from `<checksum>` to `<checksum>.module` or compress the modules before writing them to disk.

Reply:
> If we get the version, we can check much better that we don't miss an update. However, we might want to keep control over the version number since it can include our own format changes as well. E.g. we could rename the file names from `<checksum>` to `<checksum>.module` or compress the modules before writing them to disk.

When wasmer module format version is published, we can consider that plus our own version number, either in a subdir or with a minor / patch format, i.e. `{}/{}` or `{}.{}` / `{}-{}`.

Reply:
In https://github.com/CosmWasm/cosmwasm/pull/1224/commits/9f5716c00dc1ecbac82b9922f47cad71e636665b you see an approach that hacks into the private header information. This way we ensure a module version change is noticed.

We could hash the full Wasmer module header into the file path, like `format!("v3-wasmer{}", &header_hash[0..8])`. Then the bumping happens automatically.

Reply:
> In [9f5716c](https://github.com/CosmWasm/cosmwasm/commit/9f5716c00dc1ecbac82b9922f47cad71e636665b) you see an approach that hacks into the private header information. This way we ensure a module version change is noticed.
> 
> We could hash the full Wasmer module header into the file path, like `format!("v3-wasmer{}", &header_hash[0..8])`. Then the bumping happens automatically.

I like it. At least, until there's a public wasmer module format version number.

> 
> The only issue I see with this approach is that it is not obvious for admins which cache directory is used and which one is obsolete.

Not a big issue. The one being used will always contain more recent files.

Reply:
Good stuff. Let's continue the conversation in https://github.com/CosmWasm/cosmwasm/pull/1224 as Wasmer 2.2.0 will be the first version in which the relevant information is stored in the module headers.